x/bridge/observer/observer.go

```diff
@@ -0,0 +1,101 @@
+package keeper
+
+import (
+	"context"
+
+	abcitypes "github.com/cometbft/cometbft/abci/types"
+	"github.com/cometbft/cometbft/libs/log"
+	"github.com/cometbft/cometbft/libs/pubsub/query"
+	rpchttp "github.com/cometbft/cometbft/rpc/client/http"
+	coretypes "github.com/cometbft/cometbft/rpc/core/types"
+	comettypes "github.com/cometbft/cometbft/types"
+)
+
+type Observer struct {
+	logger        log.Logger
+	tmRpc         *rpchttp.HTTP
+	stopChan      chan struct{}
+	eventsOutChan chan<- abcitypes.Event
+}
+
+func NewObserver(logger log.Logger, rpcUrl string, eventsOutChan chan<- abcitypes.Event) (Observer, error) {
+	rpc, err := rpchttp.New(rpcUrl, "/websocket")
+	if err != nil {
+		return Observer{}, err
+	}
+
+	return Observer{
+		logger:        logger,
+		tmRpc:         rpc,
+		stopChan:      make(chan struct{}),
+		eventsOutChan: eventsOutChan,
+	}, nil
+}
+
+func (o *Observer) Start(ctx context.Context, queryStr string, observeEvents []string) error {
+	err := o.tmRpc.Start()
+	if err != nil {
+		o.logger.Error("Observer failed to start RPC client", err.Error())
+		return err
+	}
+
+	query, err := query.New(queryStr)
+	if err != nil {
+		o.logger.Error("Invalid query expression", err.Error())
+		return err
+	}
+
+	txs, err := o.tmRpc.Subscribe(ctx, "observer", query.String())
+	if err != nil {
+		o.logger.Error("Observer failed to subscribe to RPC client", err.Error())
+		return err
+	}
+
+	o.logger.Info("Observer starting listening for events at RPC", o.tmRpc.Remote())
+	go o.processEvents(ctx, txs, observeEvents)
+
+	return nil
+}
+
+func (o *Observer) Stop(ctx context.Context) error {
+	if err := o.tmRpc.UnsubscribeAll(ctx, "observer"); err != nil {
+		o.logger.Error("Observer failed to unsubscribe from RPC client", err.Error())
+		return err
+	}
+	close(o.stopChan)
```

Comment:
It doesn't make a lot of sense in our case (since UnsubscribeAll doesn't close the txs channel), but generally we must put the close before UnsubscribeAll to let processEvents finish before we call UnsubscribeAll. In the general case, if UnsubscribeAll closes the underlying channels, then we will get extra dirty reads from the closed channel until we close stopChan.